client/app/services/dashboard.js

```diff
@@ -1,8 +1,33 @@
-import { map, isArray } from 'underscore';
+import { map, flatten, extend, isArray } from 'underscore';
 
-function Dashboard($resource, $http, currentUser, Widget) {
+function Dashboard($resource, $http, currentUser, Widget, dashboardGridOptions) {
   function transformSingle(dashboard) {
-    dashboard.widgets = map(dashboard.widgets, row => row.map(widget => new Widget(widget)));
+    if (
+      isArray(dashboard.widgets) && (dashboard.widgets.length > 0) &&
+      isArray(dashboard.widgets[0])
+    ) {
+      // Dashboard v1 processing
+      // v1 dashboard has two columns, and widget can occupy one of them or both;
+      // this means, that there can be at most two widgets per row.
+      // Here we will map gridster columns and rows to v1-style grid
+      const dashboardV1ColumnSize = Math.round(dashboardGridOptions.columns / 2);
+      dashboard.widgets = map(
+        dashboard.widgets,
+        (row, rowIndex) => map(row, (widget, widgetIndex) => {
+          widget.options = widget.options || {};
+          widget.options.position = extend({}, {
+            row: rowIndex,
+            col: widgetIndex * dashboardV1ColumnSize,
+            sizeX: dashboardV1ColumnSize * widget.width,
+            sizeY: -1, // auto
+          }, widget.options.position);
+          return widget;
+        }),
+      );
+    }
```

Comment:
I wonder if we should do the migration here or just do it once as a database migration. 🤔

Reply:
It's up to you. This code works as expected, it's also simple and fast, but it exists only to support legacy data structures. So if it's safe to perform migration at once - we can remove this.

Reply:
OK, I will think about it some more and reach a conclusion before we finalize this.